### PR TITLE
Pipeline: simple API for assisted generation

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Assisted generation + pipelines has a bad UX at the moment: the user must manually load the assistant model (and the assistant tokenizer, if applicable), which defeats the point of being simple to use.

This PR adds the ability to specify an assistant checkpoint at pipeline definition time -- the pipeline will take care of the rest 🤗 

⚠️ While the feature was added for all pipelines that call `.generate()`, I haven't added a test on all of them. Many pipelines don't forward kwargs properly to `.generate()` which makes testing this transparent [same output, similar runtime] feature hard -- the best way to confirm assisted generation is running is by passing incompatible flags to `.generate()` to make it crash.

Example usage
```py
from transformers import pipeline
import torch

pipe = pipeline(
    "text-generation",
    model="meta-llama/Llama-3.1-8B",
    assistant_model="meta-llama/Llama-3.2-1B",  # This extra line is all that's needed!
    torch_dtype=torch.bfloat16
)
pipe_output = pipe("Once upon a time, ", max_new_tokens=50, do_sample=False)
print(pipe_output[0]["generated_text"])
```

### CODE REVIEW

File: docs/source/en/generation_strategies.md [hunk 1]
```diff
@@ -416,16 +416,6 @@ Assisted decoding assumes the main and assistant models have the same tokenizer,
 Currently, only greedy search and sampling are supported with assisted decoding, and assisted decoding doesn't support batched inputs.
 To learn more about assisted decoding, check [this blog post](https://huggingface.co/blog/assisted-generation).
 
-#### Universal Assisted Decoding
```
Comment (gante) [inline docs/source/en/generation_strategies.md:419]: The current version of the docs had the normal assisted generation docs under `Universal Assisted Decoding` [modification to the original technique to support different tokenizers]

Most of the diff is to isolate `Universal Assisted Decoding`

File: src/transformers/generation/flax_utils.py [hunk 2]
```diff
@@ -347,7 +347,6 @@ def generate(
             eos_token_id = generation_config.eos_token_id
             if isinstance(eos_token_id, list):
                 eos_token_id = eos_token_id[0]
-            logger.warning(f"Setting `pad_token_id` to `eos_token_id`:{eos_token_id} for open-end generation.")
```
Comment (gante) [inline src/transformers/generation/flax_utils.py:350]: This is a legacy warning (before I joined 👀 ) that 
a) is noisy 
b) doesn't affect generation other than the ability to try to infer the attention mask when it is not passed

In the specific case of assisted generation, it was emitted once per assistant model call, so many times 😅 More harmful than useful.

**Comment by LysandreJik** [comment]
Thanks for the nice PR @gante! 

@Rocketknight1, can you give it a first look as the pipeline owner? Thanks!

File: src/transformers/generation/flax_utils.py (hunk 2 above)
Comment (Rocketknight1) [reply src/transformers/generation/flax_utils.py:350]: Strong agree on removing this one!

File: src/transformers/pipelines/text_generation.py [hunk 3]
```diff
@@ -192,12 +191,13 @@ def _sanitize_parameters(
 
         if stop_sequence is not None:
             stop_sequence_ids = self.tokenizer.encode(stop_sequence, add_special_tokens=False)
-            if len(stop_sequence_ids) > 1:
-                warnings.warn(
-                    "Stopping on a multiple token sequence is not yet supported on transformers. The first token of"
-                    " the stop sequence will be used as the stop sequence string in the interim."
-                )
-            generate_kwargs["eos_token_id"] = stop_sequence_ids[0]
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/pipelines/text_generation.py:200]: Do we support this for sequences of IDs as well as stop strings?

File: src/transformers/pipelines/text_to_audio.py [hunk 4]
```diff
@@ -148,10 +148,9 @@ def _forward(self, model_inputs, **kwargs):
         else:
             if len(generate_kwargs):
                 raise ValueError(
-                    f"""You're using the `TextToAudioPipeline` with a forward-only model, but `generate_kwargs` is non empty.
-                                 For forward-only TTA models, please use `forward_params` instead of of
-                                 `generate_kwargs`. For reference, here are the `generate_kwargs` used here:
-                                 {generate_kwargs.keys()}"""
+                    "You're using the `TextToAudioPipeline` with a forward-only model, but `generate_kwargs` is non "
+                    "empty. For forward-only TTA models, please use `forward_params` instead of `generate_kwargs`. "
+                    f"For reference, here are the `generate_kwargs` used here: {generate_kwargs.keys()}"
```
Comment (Rocketknight1) [inline src/transformers/pipelines/text_to_audio.py:153]:
```suggestion
                    f"For reference, the `generate_kwargs` used here are: {generate_kwargs.keys()}"
```
Tiny style nit to avoid repeating "here"

File: src/transformers/pipelines/base.py [hunk 5]
```diff
@@ -425,6 +425,56 @@ def get_default_model_and_revision(
...
+        Tuple: The loaded assistant model and (optionally) the loaded tokenizer.
+    """
+    if not model.can_generate() or assistant_model is None:
+        return None, None
+
+    # If the model is passed as a string, load the model and the corresponding tokenizer
+    if isinstance(assistant_model, str):
+        assistant_config = AutoConfig.from_pretrained(assistant_model)
+        _, loaded_assistant_model = infer_framework_load_model(assistant_model, config=assistant_config)
+        loaded_assistant_model = loaded_assistant_model.to(device=model.device, dtype=model.dtype)
```
Comment (Rocketknight1) [inline src/transformers/pipelines/base.py:454]: `infer_framework_load_model` is framework-agnostic code, but `.to()` is PyTorch-only. Maybe we should add a warning if a user tries to use this with TF, since it's not supported at all?

**Review by Rocketknight1 — APPROVED** [review]
This seems very clean overall! The changes to pipeline code are relatively small, because all the actual action happens in `generate()`, so this PR really just takes care of forwarding assistant models/tokenizers to that method.

**Comment by danielkorat** [comment]
@gante will this PR make it to next release? 🙏

**Comment by danielkorat** [comment]
Hi @LysandreJik 👋 
Can we try to merge this before the next release?
This will drastically shorten the code examples we used in the HF blog and social posts.
Thanks!

**Comment by danielkorat** [comment]
hi @gante, could you please merge when free? 🙏

File: src/transformers/pipelines/text_generation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/pipelines/text_generation.py:200]: Yes, `generate` supports using multiple token IDs as stopping criteria! 

In this particular case, this warning predates the introduction of multiple token IDs as stopping criteria :) So the warning is no longer true (and hasn't been for a while)

File: tests/pipelines/test_pipelines_text_generation.py [hunk 6]
```diff
@@ -609,3 +609,17 @@ def test_pipeline_length_setting_warning(self):
         with CaptureLogger(logger) as cl:
             _ = text_generator(prompt, max_length=10)
         self.assertNotIn(logger_msg, cl.out)
+
+    @require_torch
+    def test_pipeline_assisted_generation(self):
+        """Tests that we can run assisted generation in the pipeline"""
+        model = "distilbert/distilgpt2"
```
[the remark this answers is not in the document]
Comment (gante) [reply tests/pipelines/test_pipelines_text_generation.py]: done 👍 (and confirmed that it works)

File: src/transformers/pipelines/base.py (hunk 5 above)
Comment (gante) [reply src/transformers/pipelines/base.py:454]: Added the following after checking that `assistant_model` is not `None`, and before the commented lines. In a nutshell, the cases where `.to` is not supported are caught in advance

```py
if not isinstance(model, PreTrainedModel):
        raise ValueError(
            "Assisted generation, triggered by the `assistant_model` argument, is only available for "
            "`PreTrainedModel` model instances. For instance, TF or JAX models are not supported."
        )
```

Comment (Rocketknight1) [reply src/transformers/pipelines/base.py:460]: Yes, looks good!

**Comment by Rocketknight1** [comment]
@gante yes, you can go ahead and merge this whenever you're happy!
